src/components/modals/login/utils.tsx

```diff
@@ -0,0 +1,108 @@
+import { useLinkEvmAddress } from '@/services/subsocial/evmAddresses/mutation'
+import { getSubsocialApi } from '@/subsocial-query/subsocial/connection'
+import { decodeAddress } from '@polkadot/keyring'
+import { BN, u8aToHex } from '@polkadot/util'
+import { useEffect, useState } from 'react'
+import { useDisconnect, useSignMessage } from 'wagmi'
+
+export const buildMsgParams = async (substrateAddress: string) => {
+  const decodedAddress = u8aToHex(decodeAddress(substrateAddress))
+  const subsocialApi = await getSubsocialApi()
+
+  const api = await subsocialApi.blockchain.api
+
+  const account = await api.query.system.account(substrateAddress)
+
+  const nonce = account.nonce.add(new BN(1)).toString()
+
+  return `Link to Subsocial address ${decodedAddress.replace(
+    '0x',
+    ''
+  )} (in hex) with nonce ${nonce}`
+}
+
+export const useSignEvmLinkMessage = () => {
+  const { signMessageAsync } = useSignMessage()
+  const [isSigningMessage, setIsSigningMessage] = useState(false)
+  const [isError, setIsError] = useState(false)
+
+  const signEvmLinkMessage = async (
+    emvAddress?: string,
+    substrateAddress?: string | null
+  ) => {
+    setIsError(false)
+
+    if (!emvAddress || !substrateAddress) return
+
+    const message = await buildMsgParams(substrateAddress)
+    try {
+      setIsSigningMessage(true)
+      const data = await signMessageAsync({ message })
+
+      setIsSigningMessage(false)
+      return data.toString()
+    } catch {
+      setIsError(true)
+      setIsSigningMessage(false)
+      return
+    }
+  }
+
+  return { signEvmLinkMessage, isSigningMessage, isError }
+}
+
+type SignMessageAndLinkAddressProps = {
+  setModalStep?: () => void
+  onError?: () => void
+  linkedEvmAddress?: string
+}
+
+export const useSignMessageAndLinkEvmAddress = ({
```

Comment:
I saw that you used this hook in multiple places, outside of the this login folder, so I think it might be better to move this to the `hooks/` folder

because what I think when I see utils inside of `login` folder, it should only be used internally inside components in that `login` folder